staging/src/k8s.io/kubectl/pkg/cmd/exec/exec.go

```
@@ -208,7 +209,8 @@ func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []s
 	if len(argsIn) > 0 && argsLenAtDash != 0 {
 		p.ResourceName = argsIn[0]
 	}
-	if argsLenAtDash > -1 {
+	// we expect argsLenAtDash = 1
+	if argsLenAtDash == 0 || argsLenAtDash == 1 {
 		p.Command = argsIn[argsLenAtDash:]
```

I think this should be merged with the previous condition which checks that `argsIn` is non-empty. If `argsIn` is empty here and `argsLenAtDash` is 1, `argsIn[1:]` throws a slice out of range.

---

https://github.com/cubefs/cubefs/blob/f4c9dac980752ef4ccaefdca2e923a2a9202716a/depends/spf13/pflag/flag.go#L1092

pflag guarantees `argsLenAtDash <= len(args)`. Do you think it is necessary to add the range check?

---

I see. Thanks. Perhaps adding a note on the function about this will be helpful for anyone seeing the code for the first time.

With that said, I don't quite understand the `len(argsIn) > 1` condition in the `else if`. That's almost a guaranteed outcome because as you said `argsLenAtDash <= len(argsIn)`. This condition would only be considered when `argsLenAtDash` >= 2 which implies `argsIn` >= 2.

---

no.

Let me put together a table to make it clearer.

| command | argsIn | argsLenAtDash | the command will be execute right now | 
| -- | -- | -- | -- |
| `kubectl exec -it pod-0 -- bash run.sh foo` | [bash run.sh foo] | 1 | bash run.sh foo ✅ |
| `kubectl exec -it pod-0 -- bash -- run.sh foo` | [bash -- run.sh foo] | 1 | bash -- run.sh foo ✅ |
| `kubectl exec -it pod-0 bash -- run.sh foo` | [bash run.sh foo] | 2 | run.sh foo ❌ |
| `kubectl exec -it pod-0 bash bar -- run.sh foo` | [bash bar run.sh foo] | 3 | run.sh foo ❌ |

In simple terms, argsLenAtDash is the position of the first `--`, and the first `--` will be removed from args.

---

Well, yes, I think I got that. Thanks. My question was with the `len(argsIn) > 1` condition. At that point we know that to be true as implied by the condition `argsLenAtDash` <= `len(argsIn)`, so if I'm not mistaken, there's no point of keeping that logical predicate. 

How I understand it is that if you are executing a command in a pod, we are making it so that it is required that there be *exactly* 1 arg representing the name of the pod/resource before the execution script delimited by the (very first) `--`. In your code, however, there's branching of conditions, which I think would be simplified by just returning an error when `argsLenAtDash != 1`.

```go
func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []string, argsLenAtDash int) error {
    if argsLenAtDash != 1 {
        		return cmdutil.UsageErrorf(cmd, "exec [POD] [COMMAND] is not supported anymore. Use exec [POD] -- [COMMAND] instead")

    }
//...
}
```

---

if argsLenAtDash == 0, command will be `kubectl exec -- echo 123`, should receive `error: pod, type/name or --filename must be specified`. This is another error and should not be returned here.

---

Based on that error, is the following command execution valid?

```
kubectl exec --filename path/to/file.name -- echo 123
```

i.e. it's ok to not have an pod name unless there's a filename? Apologies as I've never used kubectl exec this way. From your code it also looks like `--filename` option cannot be used when a pod is specified correct?

I think my recommendation in general is to fail early by listing down all unacceptable conditions first and then returning the relevant errors. Something like:

```go
func (p *ExecOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, argsIn []string, argsLenAtDash int) error {
    if unacceptableCase1 {
        // return error here
    } else if unacceptableCase2 {
        // return error here
    } else if unacceptableCaseN {
        // return error here
    }
// only continue processing the args when we know that exec is structured correctly.
}
```

I think this way, it's easier to read because it tells you outright which conditions are not allowed.

---

Validate is in https://github.com/kubernetes/kubernetes/blob/b53b9fb5573323484af9a19cf3f5bfe80760abba/staging/src/k8s.io/kubectl/pkg/cmd/exec/exec.go#L248-L259

An independent Validate function is a more common practice, where the result of Complete is checked.

---

WDYT

---

I think it is right to move it to `Validate` function.

But this involves the behavior change of `the usage of cmdutil.UsageErrorf`, `cmdutil.UsageErrorf` is only used in Complete, not in Validate.

I don't know why this is, and I don't want to make changes in user behavior in this PR.

---

ping again